### DISCUSSION TITLE
Adding support for Bluelink Australia

### DISCUSSION THREAD

**Discussion by muneeb1990** [Hyundai-Kia-Connect/kia_uvo#462] [Ideas]
**Is your feature request related to a problem? Please describe.**
Hi, I recently purchased a Hyundai Palisade which is the first model to support Bluelink in Australia and would like to assist in capturing the relevant API endpoints and data to add support for Australia in this HA custom component

**Describe the solution you'd like**
Add support for Australia models

**Include the logs of the end point from Data in HA if this is a sensor**
Flow for login:
```
Flow Details
2022-11-06 13:13:50 POST https://au-apigw.ccs.hyundai.com.au:8080/api/v1/user
                         /signin
                         ← 200 OK application/json 199b 332ms
          Request                   Response                    Detail
Host:             au-apigw.ccs.hyundai.com.au:8080
Content-Type:     text/plain;charset=UTF-8
Origin:           https://au-apigw.ccs.hyundai.com.au:8080
Accept-Encoding:  gzip, deflate, br
Cookie:           account=ZDQ5Y2U4ZjUtYWVjMS00ZGI5LTljOGEtNzlmZmE4N2MwOGE3.23c
                  4ee838b095dec791cc428cc475d43ff0a03b9
Connection:       keep-alive
Accept:           */*
User-Agent:       Mozilla/5.0 (iPhone; CPU iPhone OS 16_1 like Mac OS X)
                  AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148
Referer:          https://au-apigw.ccs.hyundai.com.au:8080/web/v1/user/signin
Content-Length:   81
Accept-Language:  en-AU,en;q=0.9
Raw                                                                     [m:auto]
{"email":"email@user.com","password":"XXXXXXXXXX","mobileNum":""
}

```
Flow for getting latest status:
```
Flow Details
2022-11-06 13:20:53 GET https://au-apigw.ccs.hyundai.com.au:8080/api/v1/spa/v
                        ehicles/<98778a3a-xxxx-xxxx-yyyy-64e3cc96dxxx>/status/l
                        atest
                        ← 200 OK application/json 1.62k 176ms
          Request                   Response                    Detail
ccsp-application-id:  55e32c34-6e47-4122-ad61-c74904ef975c
Accept-Encoding:      br;q=1.0, gzip;q=0.9, deflate;q=0.8
Stamp:                Yphzo3vRdjKtSNVx4ICDzrDfrMpHivrb/Qcf6LxWuNTUWPAdDOccnYwa
                      vcbdaR0=
offset:               11
User-Agent:           AU_BlueLink/1.0.0 (com.hyundai.bluelink.aus;
                      build:10030; iOS 16.1.0) Alamofire/5.4.4
If-None-Match:        W/"67c-ij9FLxCAtRYPWvIKfeHTUNJUYvo"
Connection:           keep-alive
Content-Type:         application/json; charset=UTF-8
ccsp-device-id:       89032bfc-efcf-4fb0-aa6e-a62bb3236843
Cookie:               profile=MjA3Yjg5ZmYtZjQ4ZC00ZDcyLWFlNDMtYWQwMDI4YjhkNmJm
                      .c8bc52d2904afd9c6e70319f2bd3693b93ecd3aa; account=ZDQ5Y
                      2U4ZjUtYWVjMS00ZGI5LTljOGEtNzlmZmE4N2MwOGE3.23c4ee838b09
                      5dec791cc428cc475d43ff0a03b9
No request content                                                      [m:auto]
⇩  [260/261]                                                            [*:8881]
```
**Describe alternatives you've considered**

**Additional context**
Please let me know what else you need to get this going. Last but not the least, thanks for supporting this!

**Comment by cdnninja** [comment]
This is a great start for this.  Would you consider developer some of this with our support or guidance? It's always easier when you have access to the car.

**Comment by muneeb1990** [comment]
Yes, of course.

**Comment by fuatakgun** [comment]
i have a weak feeling that, this is identical to EU logic, just different app-id and cs-id.

**Comment by cdnninja** [comment]
> i have a weak feeling that, this is identical to EU logic, just different app-id and cs-id.

Hopefully just without stamps.

**Comment by cdnninja** [comment]
I think best way to start this one is to download postman and try manually replicate logging in and getting the vehicle data like above. 

Once successful with that we can start building this into the 2.0 version of this.  It is run via: https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api

So we would need to create a new API file that implements this: https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/blob/master/hyundai_kia_connect_api/ApiImpl.py

Examples of what it would look like are https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/blob/master/hyundai_kia_connect_api/KiaUvoApiCA.py and 
https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/blob/master/hyundai_kia_connect_api/KiaUvoApiEU.py.

Very different login schemes used between those two. 

A few other minor spots need to be updated to allow it to reference Australia but that API file is the key part. 

Feel free to drop us any questions and we will help best we can!

**Comment by muneeb1990** [comment]
Thanks, I'll tinker around with Postman over the weekend. Do I just paste them here for you guys to complete the API?

**Comment by cdnninja** [comment]
Yes or you can drop a pull request for the code changes.  I started a PR here that will reflect the work: https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/pull/115

**Comment by cdnninja** [comment]
> > i have a weak feeling that, this is identical to EU logic, just different app-id and cs-id.
> 
> Hopefully just without stamps.

Reviewing this I just saw it has stamps. I missed that.  This is over my skill.  It has been done for EU over here: https://github.com/neoPix/bluelinky-stamps

**Comment by muneeb1990** [comment]
Yeah that does seem over my skillset as well. How can I check whether these stamps can work for AU as well?

**Comment by cdnninja** [comment]
Postman I would think but knowing the exact stamp to use would be key. I would think can take the stamps from the above repo.

**Reply by grouts112** [reply]
I would love to help, but What is postman and where do I get it ?

**Reply by cdnninja** [reply]
> I would love to help, but What is postman and where do I get it ?

Appreciate the desire to support.   This is highly complex and even over my ability to direct you to solve.  Give it a google if you want to play with it.

**Comment by bitnimble** [comment]
I'm a developer and would love to contribute to this; I'm familiar with Postman etc.

I'm a little confused as to where to start though. I attempted mitmproxy / Charles on the app but unfortunately it looks like they use certificate pinning, and it _also_ looks like they verify the checksum at runtime ("This app has been forged" and then it refuses to open) so patching the app to bypass the certificate pinning doesn't work. 

@cdnninja how were you able to inspect the request/responses for eu / ca etc? is there a non-app version of Bluelink in those regions?

and @muneeb1990 how were you able to inspect those requests in your OP?

**Reply by muneeb1990** [reply]
I used MITM proxy on my iPhone. The main issue I ran into was something called ‘stamps’ that are used to authenticate Bluelink API requests. It seems like AU has a different stamp algorithm than EU and US regions and therefore the existing bluelinky-stamps is incompatible.

**Reply by muneeb1990** [reply]
Here’s the link to it: https://github.com/neoPix/bluelinky-stamps

sorry I’m on my phone and it won’t let me enter a URL.

**Reply by bitnimble** [reply]
thanks. seems like i'm gonna have to get a physical iPhone for the easiest mitm then haha

**Reply by muneeb1990** [reply]
I’m happy to send through logs if that helps. Have you looked into the stamp thing I mentioned?

**Reply by bitnimble** [reply]
yep, I can try get stamp generation working for the Australian versions but I'll just need to get a good dev environment set up first

**Reply by bitnimble** [reply]
it's all good, I have an iPhone that I can borrow for testing

**Reply by muneeb1990** [reply]
Awesome! Let me know how I can help in anyway possible.

**Reply by bitnimble** [reply]
I've joined the Discord (same username): https://discord.gg/wQT4xqj4

we can chat more there, share logs, etc

**Comment by cdnninja** [comment]
> I'm a developer and would love to contribute to this; I'm familiar with Postman etc.
> 
> 
> 
> I'm a little confused as to where to start though. I attempted mitmproxy / Charles on the app but unfortunately it looks like they use certificate pinning, and it _also_ looks like they verify the checksum at runtime ("This app has been forged" and then it refuses to open) so patching the app to bypass the certificate pinning doesn't work. 
> 
> 
> 
> @cdnninja how were you able to inspect the request/responses for eu / ca etc? is there a non-app version of Bluelink in those regions?
> 
> 
> 
> and @muneeb1990 how were you able to inspect those requests in your OP?


For Canada I just use chrome or edge developer mode as it has a website.  

EU I haven't done that's been other people.  Bluelinky discord is best spot for that.   It only has an app and isn't easy with cert pinning.

**Reply by fuatakgun** [reply]
Android is a bit harder (actually a lot for me) because rooting/cert pinning doesn't work for me anymore.

There are some users relying on iOS emulator and Charles proxy to reverse engineer ongoing traffic.

**Reply by bitnimble** [reply]
Yeah, iOS emulation was my backup option but I wasn't sure if it also had pinning of some sort (googled and it looks like that is a thing on iOS too). If it doesn't then sounds like that's much easier :)  I'm currently just decompiling and stripping out the forged checks and seeing if that works... otherwise maybe I can attach a debugger and inject the unpin at runtime.

thanks for the tip on the Discord, will join and ask there to see what info is around

**Comment by muneeb1990** [comment]
Just joined too, cdf952.

Sent from my iPhone

On 24 Jul 2023, at 10:23 pm, bitnimble ***@***.***> wrote:

﻿

I've joined the Discord (same username): https://discord.gg/wQT4xqj4

we can chat more there, share logs, etc

**Comment by fuatakgun** [comment]
if we need different stamp generation algorithm, we can work with this repository (using SO files from respective apps) to generate APP ID specific files. 

https://github.com/Hacksore/hyundai-kia-stamp/tree/master/lib/hyundai

**Comment by cdnninja** [comment]
Hey all!   This is now been added with basic support for Hyundai. Lets track issues here: https://github.com/Hyundai-Kia-Connect/kia_uvo/issues/692

**Comment by zappoo** [comment]
Any chance this will get added for Kia?

**Comment by zappoo** [comment]
I've tried to do some basic POSTS to `https://au-apigw.ccs.kia.com.au:8082/api/v1/user/signin` using a raw body like

`{"email":"email@user.com","password":"XXXXXXXXXX","mobileNum":""}`

No luck - keep getting this error:

```
{
    "errId": "6c3f7e37-d8f9-4067-8996-55bb1dd3ec09",
    "errCode": "4010",
    "errMsg": "Require authentication"
}

If we can get this working, then I am happy to invest some time in a PR for this API.

Kia Connect Aus only has an android app - so it's hard to sniff traffic sadly.

**Reply by cdnninja** [reply]
Pretty certain it follows standard approach. Needs app id and other ids.  Biggest thing is the xor value.

**Comment by zappoo** [comment]
Given there is no website and only an Android app, debugging is quite difficult :( How does the XOR value get generated?

**Comment by grouts112** [comment]
“ only an Android app”?
There is an ios app, I use it on my KIA Niro in Australia.
I am of the understanding this can be run on my mac, and therefore traced etc?
I have all the above, I just do not have the knowledge on what todo,
Happy to give it a go with some patient guidance.

**Comment by zappoo** [comment]
Yes sorry - i meant "_only a mobile app, not a website_".. as without a website it is very had to sniff traffic.

If you have it on a Mac, then that would be ideal. You would need to install a tool like "FIddler", run it whilst you play with the app on your mac and somehow share what fiddler captures.

**Reply by grouts112** [reply]
Ok, sounds like a plan, I will have a tinker when I get a chance
